File: backend/app/core/lifecycle.py
# Concept by MrHan (08974747477)
"""Issue lifecycle state machine (ADR-012).

Central definition of allowed status transitions. CLOSED -> REOPENED happens only
through the dedicated reopen endpoint; the generic status endpoint uses the
transitions declared here.
"""

from __future__ import annotations

from app.models.enums import IssueStatus

# Allowed transitions for the generic status-change endpoint.
ALLOWED_TRANSITIONS: dict[IssueStatus, set[IssueStatus]] = {
    IssueStatus.OPEN: {IssueStatus.IN_PROGRESS, IssueStatus.PENDING, IssueStatus.CLOSED},
    IssueStatus.IN_PROGRESS: {
        IssueStatus.PENDING,
        IssueStatus.CLOSED,
        IssueStatus.OPEN,
    },
    IssueStatus.PENDING: {
        IssueStatus.IN_PROGRESS,
        IssueStatus.OPEN,
        IssueStatus.CLOSED,
    },
    # CLOSED -> REOPENED only via the reopen endpoint (not the generic status one).
    IssueStatus.CLOSED: set(),
    IssueStatus.REOPENED: {
        IssueStatus.IN_PROGRESS,
        IssueStatus.PENDING,
        IssueStatus.CLOSED,
    },
}


def can_transition(current: IssueStatus, target: IssueStatus) -> bool:
    return target in ALLOWED_TRANSITIONS.get(current, set())

File: frontend/src/features/issues/lifecycle.ts
// Concept by MrHan (08974747477)
// UI mirror of the backend issue lifecycle (ADR-012) for gating action buttons.
// The backend remains authoritative — the UI only avoids offering transitions the
// server will reject.
import type { IssueStatus } from '@/api/types';

// Valid targets for the generic /status endpoint (excludes CLOSED — use Close;
// excludes any transition out of CLOSED — use Reopen).
const STATUS_TRANSITIONS: Record<IssueStatus, IssueStatus[]> = {
  OPEN: ['IN_PROGRESS', 'PENDING'],
  IN_PROGRESS: ['PENDING', 'OPEN'],
  PENDING: ['IN_PROGRESS', 'OPEN'],
  CLOSED: [],
  REOPENED: ['IN_PROGRESS', 'PENDING'],
};

export function allowedStatusTargets(current: IssueStatus): IssueStatus[] {
  return STATUS_TRANSITIONS[current] ?? [];
}

export function canClose(status: IssueStatus): boolean {
  return status !== 'CLOSED';
}

export function canReopen(status: IssueStatus): boolean {
  return status === 'CLOSED';
}

export function canEditMetadata(status: IssueStatus, isArchived: boolean): boolean {
  // Backend rejects metadata edits on CLOSED or archived issues.
  return status !== 'CLOSED' && !isArchived;
}
